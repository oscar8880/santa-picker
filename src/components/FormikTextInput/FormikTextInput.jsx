import styles from './FormikTextInput.module.scss';
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { useField } from 'formik';

const FormikTextInput = ({ label, ...props }) => {
  const [inFocus, setInFocus] = useState(false);
  const [valuePresent, setValuePresent] = useState(false);
  const [field, meta] = useField(props);

  const handleChange = (e) => {
    if (e.target.value === '') {
      setValuePresent(false);
    } else {
      setValuePresent(true);
    }
  };

  return (
    <div
      className={styles.FormikTextInput__Container}
      onFocus={() => setInFocus(true)}
      onBlur={() => setInFocus(false)}
      onChange={handleChange}
    >
      <label
        htmlFor={props.id || props.name}
        className={classNames(styles.FormikTextInput__Label, {
          [styles.FormikTextInput__LabelAsLabel]: inFocus || valuePresent,
          [styles.FormikTextInput__LabelAsPlaceholder]:
            !inFocus && !valuePresent,
        })}
      >
        {label}
      </label>
      <input className={styles.FormikTextInput__Input} {...field} {...props} />
      {meta.touched && meta.error ? (
        <div className={styles.FormikTextInput__Error}>{meta.error}</div>
      ) : null}
    </div>
  );
};

FormikTextInput.propTypes = {
  label: PropTypes.string,
  name: PropTypes.string,
  id: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
};

export default FormikTextInput;
