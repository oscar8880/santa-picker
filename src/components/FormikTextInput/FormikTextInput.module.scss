@import "../../scss/functions";
@import "../../scss/variables";

.FormikTextInput__Container {
  position: relative;
}

.FormikTextInput__Input {
  border: 1px solid $grey-300;
  border-radius: 4px;
  padding-bottom: $spacer-4;
  padding-left: $spacer-4;
  padding-top: $spacer-4;
  width: 100%;
}

.FormikTextInput__Label {
  color: $grey-600;
  position: absolute;
  transition: $transition-default;
}

.FormikTextInput__LabelAsLabel {
  background-color: $white;
  color: $grey-600;
  font-size: $font-size-small;
  font-weight: bolder;
  left: $spacer-4;
  padding-left: $spacer-1;
  padding-right: $spacer-1;
  top: -$spacer-2 - 2px;
}

.FormikTextInput__LabelAsPlaceholder {
  color: $grey-600;
  left: $spacer-4;
  position: absolute;
  top: $spacer-4;
}

.FormikTextInput__Error {
  color: $danger-colour;
  font-size: $font-size-xsmall;
  font-weight: bolder;
  padding-left: $spacer-4;
  padding-top: $spacer-1;
  position: absolute;
}
