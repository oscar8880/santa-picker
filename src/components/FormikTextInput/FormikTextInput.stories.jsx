import React from 'react';
import FormikTextInput from './FormikTextInput';
import { Formik, Form } from 'formik';

export default {
  component: FormikTextInput,
  title: 'FormikTextInput',
};

export const Default = () => {
  return (
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <FormikTextInput name="input" id="input" label="Input" />
      </Form>
    </Formik>
  );
};
