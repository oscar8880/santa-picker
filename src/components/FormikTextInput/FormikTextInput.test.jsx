import React from 'react';
import { render } from '@testing-library/react';
import { Formik, Form } from 'formik';

import FormikTextInput from './FormikTextInput';

test('matches snapshot', async () => {
  const { asFragment } = render(
    <Formik initialValues={{}} onSubmit={() => {}}>
      <Form>
        <FormikTextInput name="input" id="input" label="Input" />
      </Form>
    </Formik>,
  );

  expect(asFragment()).toMatchSnapshot();
});
