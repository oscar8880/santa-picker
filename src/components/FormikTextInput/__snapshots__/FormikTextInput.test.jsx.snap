// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="FormikTextInput__Container"
    >
      <label
        class="FormikTextInput__Label FormikTextInput__LabelAsPlaceholder"
        for="input"
      >
        Input
      </label>
      <input
        class="FormikTextInput__Input"
        id="input"
        name="input"
        value=""
      />
    </div>
  </form>
</DocumentFragment>
`;
